drivers/bluetooth/hci/cyw43xxx.c

```diff
@@ -0,0 +1,300 @@
+
+	/* Copy data */
+	(void)memcpy(data_buf, (void *)hci_data, HCI_VSC_UPDATE_BAUD_RATE_LENGTH);
+
+	/* Send update uart baudrate command. */
+	err = bt_hci_cmd_send_sync(BT_HCI_VND_OP_UPDATE_BAUDRATE, buf, NULL);
+	if (err) {
+		return err;
+	}
+
+	/* Re-configure Uart baudrate */
+	err = bt_hci_uart_set_baudrate(bt_uart_dev, baudrate);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+static int bt_firmware_download(const uint8_t *firmware_image, uint32_t size)
+{
+	uint8_t *data = (uint8_t *) firmware_image;
+	volatile uint32_t remaining_length = size;
+	struct net_buf *buf;
+	int err;
+
+	LOG_DBG("Executing Fw downloading for CYW43xx device");
+
+	/* Send hci_download_minidriver command */
+	err = bt_hci_cmd_send_sync(BT_HCI_VND_OP_DOWNLOAD_MINIDRIVER, NULL, NULL);
+	if (err) {
+		return err;
+	}
+
+	/* The firmware image (.hcd format) contains a collection of hci_write_ram
+	 * command + a block of the image, followed by a hci_write_ram image at the end.
+	 * Parse and send each individual command and wait for the response. This is to
+	 * ensure the integrity of the firmware image sent to the bluetooth chip.
+	 */
+	while (remaining_length) {
+		size_t data_length = data[2]; /* data length from firmware image block */
+		uint16_t op_code = *(uint16_t *) data;
+
+		/* Allocate buffer for hci_write_ram/hci_launch_ram command. */
+		buf = bt_hci_cmd_create(op_code, data_length);
+		if (buf == NULL) {
+			printk("Unable to allocate command buffer\n");
+			return err;
+		}
+
+		/* Add data part of packet */
+		void *data_buf = net_buf_add(buf, data_length);
+
+		(void)memcpy(data_buf, (void *)&data[3], data_length);
+
+		/* Send hci_write_ram command. */
+		err = bt_hci_cmd_send_sync(op_code, buf, NULL);
+		if (err) {
+			return err;
+		}
+
+		switch (op_code) {
+		case BT_HCI_VND_OP_WRITE_RAM:
+			/* Update remaining length and data pointer:
+			 * content of data length + 2 bytes of opcode and 1 byte of data length.
+			 */
+			data += data_length + 3;
+			remaining_length -= data_length + 3;
+			break;
+
+		case BT_HCI_VND_OP_LAUNCH_RAM:
+			remaining_length = 0;
+			break;
+
+		default:
+			return -1;
+		}
+	}
+
+	LOG_DBG("Fw downloading complete");
+	return 0;
+}
+
+int bt_h4_vnd_setup(const struct device *dev)
+{
+	uint32_t baudrate_for_fw_download = DT_PROP(HCI_UART_NODE, baudrate_for_fw_download);
+	uint32_t baudrate_for_operation = DT_PROP(DT_CHOSEN(zephyr_bt_uart), current_speed);
+	struct gpio_dt_spec bt_reg_on = GPIO_DT_SPEC_GET(HCI_UART_NODE, bt_reg_on_gpios);
+	int err;
+
+	/* Check BT Uart instance */
+	if (!device_is_ready(dev)) {
+		return -EINVAL;
+	}
+
+	/* Check BT REG_ON gpio instance */
+	if (!device_is_ready(bt_reg_on.port)) {
+		printk("Error: failed to configure bt_reg_on %s pin %d\n",
+		       bt_reg_on.port->name, bt_reg_on.pin);
+		return -EIO;
+	}
+
+	/* Configure bt_reg_on as output  */
+	err = gpio_pin_configure_dt(&bt_reg_on, GPIO_OUTPUT | GPIO_PULL_UP);
```

Comment:
I think `GPIO_PULL_UP` should be dropped. First, pull-up/down confguration normally comes from the devicetree flags and they are applied automatically by using the `_dt` api, but also what's the point since you are setting it as an output?